sys/kern/kern_exec.c

```diff
@@ -1153,15 +1157,78 @@ exec_new_vmspace(struct image_params *imgp, struct sysentvec *sv)
 	} else {
 		ssiz = maxssiz;
 	}
+#if __has_feature(capabilities)
+	if ((sv->sv_flags & SV_CHERI) != 0) {
+		/*
+		 * NB: This may cause the stack to exceed the administrator-
+		 * configured stack size limit.
+		 */
+		ssiz = CHERI_REPRESENTABLE_LENGTH(ssiz);
+	}
+#endif
 	imgp->eff_stack_sz = lim_cur(curthread, RLIMIT_STACK);
 	if (ssiz < imgp->eff_stack_sz)
 		imgp->eff_stack_sz = ssiz;
-	stack_addr = sv->sv_usrstack - ssiz;
+	p->p_usrstack = sv->sv_usrstack;
+#if __has_feature(capabilities)
+	if ((sv->sv_flags & SV_CHERI) != 0)
+		p->p_usrstack = CHERI_REPRESENTABLE_BASE(p->p_usrstack, ssiz);
+#endif
+
+	stack_addr =  p->p_usrstack - ssiz;
+	imgp->stack_sz = ssiz;
 	error = vm_map_stack(map, stack_addr, (vm_size_t)ssiz,
 	    obj != NULL && imgp->stack_prot != 0 ? imgp->stack_prot :
 	    sv->sv_stackprot, VM_PROT_ALL, MAP_STACK_GROWS_DOWN);
 	if (error != KERN_SUCCESS)
 		return (vm_mmap_to_errno(error));
+#if __has_feature(capabilities)
+	/*
+	 * Our capability must cover the whole stack, so roundup the
+	 * length to be representable even if it means the capability
+	 * extends beyond the stack bounds.
+	 *
+	 * XXX: Should we instead unconditionally adjust the actual
+	 * stack location?
+	 */
+	stack_offset = 0;
+	do {
+		rounded_stack_addr = CHERI_REPRESENTABLE_BASE(
+		    stack_addr, ssiz + stack_offset);
+		stack_offset = stack_addr - rounded_stack_addr;
+	} while (rounded_stack_addr != CHERI_REPRESENTABLE_BASE(
```

Comment:
Do we still need this? We could round to representable base `stack_addr` and update `p_usrstack = rounded_stack_addr + ssiz` and then map `CHERI_REPRESENTABLE_LENGTH(ssiz)` from `stack_addr`. This would avoid the loop to make the capability representable, and will be a single `vm_map_reservation_create()` call with the reservation patch.

Reply:
I waffled on this a bit. I think that for all current CHERI architectures, it's fine to move the stack down as is convenient so maybe we should just make this unconditional above and drop this (without the part of the change the kernel panics on hybrid code). I fear we might face some pushback on x86, but likely we should leave that worry for the future and add x86 specific code if it's a problem.